servers/rendering/renderer_rd/shaders/volumetric_fog.glsl

```diff
@@ -239,465 +108,156 @@ const vec3 halton_map[TEMPORAL_FRAMES] = vec3[](
-		if (all(bvec2(all(greaterThanEqual(position, vec3(0.0))), all(lessThan(position, voxel_gi_instances.data[i].bounds))))) {
-			position /= voxel_gi_instances.data[i].bounds;
-
-			vec4 light = vec4(0.0);
-			for (uint j = 0; j < voxel_gi_instances.data[i].mipmaps; j++) {
-				vec4 slight = textureLod(sampler3D(voxel_gi_textures[i], linear_sampler_with_mipmaps), position, float(j));
-				float a = (1.0 - light.a);
-				light += a * slight;
-			}
-
-			light.rgb *= voxel_gi_instances.data[i].dynamic_range * params.gi_inject;
-
-			total_light += light.rgb;
+	if (cull_mask > 0.0) {
+		{
+#CODE : FOG
 		}
-	}
-
-	//sdfgi
-#ifdef ENABLE_SDFGI
-
-	{
-		float blend = -1.0;
-		vec3 ambient_total = vec3(0.0);
-
-		for (uint i = 0; i < sdfgi.max_cascades; i++) {
-			vec3 cascade_pos = (world_pos - sdfgi.cascades[i].position) * sdfgi.cascades[i].to_probe;
 
-			if (any(lessThan(cascade_pos, vec3(0.0))) || any(greaterThanEqual(cascade_pos, sdfgi.cascade_probe_size))) {
-				continue; //skip cascade
-			}
-
-			vec3 base_pos = floor(cascade_pos);
-			ivec3 probe_base_pos = ivec3(base_pos);
-
-			vec4 ambient_accum = vec4(0.0);
-
-			ivec3 tex_pos = ivec3(probe_base_pos.xy, int(i));
-			tex_pos.x += probe_base_pos.z * sdfgi.probe_axis_size;
-
-			for (uint j = 0; j < 8; j++) {
-				ivec3 offset = (ivec3(j) >> ivec3(0, 1, 2)) & ivec3(1, 1, 1);
-				ivec3 probe_posi = probe_base_pos;
-				probe_posi += offset;
-
-				// Compute weight
-
-				vec3 probe_pos = vec3(probe_posi);
-				vec3 probe_to_pos = cascade_pos - probe_pos;
-
-				vec3 trilinear = vec3(1.0) - abs(probe_to_pos);
-				float weight = trilinear.x * trilinear.y * trilinear.z;
-
-				// Compute lightprobe occlusion
-
-				if (sdfgi.use_occlusion) {
-					ivec3 occ_indexv = abs((sdfgi.cascades[i].probe_world_offset + probe_posi) & ivec3(1, 1, 1)) * ivec3(1, 2, 4);
-					vec4 occ_mask = mix(vec4(0.0), vec4(1.0), equal(ivec4(occ_indexv.x | occ_indexv.y), ivec4(0, 1, 2, 3)));
-
-					vec3 occ_pos = clamp(cascade_pos, probe_pos - sdfgi.occlusion_clamp, probe_pos + sdfgi.occlusion_clamp) * sdfgi.probe_to_uvw;
-					occ_pos.z += float(i);
-					if (occ_indexv.z != 0) { //z bit is on, means index is >=4, so make it switch to the other half of textures
-						occ_pos.x += 1.0;
-					}
-
-					occ_pos *= sdfgi.occlusion_renormalize;
-					float occlusion = dot(textureLod(sampler3D(sdfgi_occlusion_texture, linear_sampler), occ_pos, 0.0), occ_mask);
-
-					weight *= max(occlusion, 0.01);
+#ifdef DENSITY_USED
+		density *= cull_mask;
+		if (abs(density) > 0.001) {
+			int final_density = int(density * DENSITY_SCALE);
+			imageAtomicAdd(density_only_map, pos, uint(final_density));
+
+#ifdef EMISSION_USED
+			{
+				emission *= clamp(density, 0.0, 1.0);
+				emission = clamp(emission, vec3(0.0), vec3(8.0));
+				// Convert to UNORM and scale values towards 0 to increase precision in low values
+				emission /= vec3(8.0);
+				emission = pow(emission, vec3(1.0 / 2.2));
```

Comment:
Something I just realized is that this nonlinear scaled emission that I suggested is not mathematically correct when using addition. (e.g. 0.5+0.9=1.4 becomes 0.5+0.9 ~= 3.1.) As a result, overlapping emissive volumes look far too bright. For accurate addition, you'd have to load, unscale, add, rescale, then store. There's no reasonable way this is possible with atomic image operations.

The alternative reduced-bits option is unfortunate because 7 bits per color yields significant color inaccuracy at low fog densities. Setting fog density below 1/127, instead of yellowing, would truncate all emission. This makes thin clouds look terrifying if you ask me.

The only ways around the imprecision problem that I can think of are:
A. Reduce the emission range
B. Add another image3D to provide more bits
C. Add a fog-volume-density linear-scale property to the environment, replacing the current nonlinear scaling. It scales the density/emission before writing and unscales it after reading, just like the nonlinear scaling. This would effectively give the user control over the range of precision.

I think option C is the best one, but that could be done in another PR.

Reply:
Ope, sorry, I moved this comment here because the other thread was marked resolved

Reply:
I added another option to the previous comments:

> D. Use emission_energy in the environment to scale all emissive colors and force emission in the shaders to cap at 1 (e.g. we would be defining emission color in the volume and emission energy globally)

Ultimately, I think going with A is probably fine. Density is usually quite low and since Emission scales with Density it will be quite low as well. In cases where both emission and density are high, clamping to 4 should be sufficient (only the emission at the edge of the volume will be seen anyway). The range of 0-8 was arbitrarily chosen, so I don't feel bad changing it. 

If we still notice issues stemming from low precision we can select one of the other options.

Reply:
Ah, I wasn't sure about the consequences of modifying the total emission range. Makes sense to me.

Reply:
With emission range reduced to 0-4, I'm still noticing color inaccuracy.

Here's an example with a box volume with dimensions (50, 25, 25), density 0.015, and white emission fading along the x-axis. Unit sized icon.png in the middle for scale.
<details><summary>Screenshot:</summary>

![image](https://user-images.githubusercontent.com/6376721/135900294-eda134c6-d74f-43a4-9357-56b41281a259.png)

</details>

No matter how many bits are added, I could keep increasing the size of the scene, fog_length, and decrease density until this issue shows again. This is why option C seemed ideal to me. Though, at some point a scene is just unreasonably big.

If I'm understanding correctly, option D seems like it would resolve the precision issue. Only issue I see is in the case of modifying the global emission_energy to allow a high emission volume in a scene. This requires scaling back emission on all other volumes individually. This might also make volumes reused in different scenes/environments visually inconsistent.

I'm worried I might be nitpicking a bit, so disregard as you please. If these points become genuine problems in the future, fixing them won't be difficult.